### ISSUE TITLE
Extmark with `number_hl_group` does not combine highlighting with any `*LineNr*` highlight group

### ISSUE THREAD

**Issue by echasnovski** [neovim/neovim#27555]
### Problem

Extmark with `number_hl_group` does not combine highlighting attributes with `*LineNr*` groups (`LineNr`, `LineNrAbove`, `LineNrBelow`, `CursorLineNr`).

### Steps to reproduce

1. Create an 'init.lua' with the following content:
    ```lua
    vim.o.termguicolors, vim.o.number, vim.o.cursorline = true, true, true

    vim.api.nvim_set_hl(0, 'Red', { fg = 'red' })
    vim.api.nvim_set_hl(0, 'CursorLineNr', { bold = true })
    vim.api.nvim_set_hl(0, 'LineNr', { reverse = true })
    vim.api.nvim_set_hl(0, 'LineNrAbove', { underline = true })
    vim.api.nvim_set_hl(0, 'LineNrBelow', { undercurl = true })

    vim.api.nvim_buf_set_lines(0, 0, -1, false, { 'Hello', 'World', 'People' })
    local ns_id = vim.api.nvim_create_namespace('tmp')
    vim.api.nvim_buf_set_extmark(0, ns_id, 1, 0, { number_hl_group = 'Red' })
    ```

1. `nvim --clean -u init.lua`. Observe that line number 2 is highlighted as plain red foreground and not reversed (as `LineNr`).
2. Navigate to line 2. Observe that line number 2 is highlighted as plain red and not bold red.
3. `:set relativenumber` and move cursor up and down. Line number 2 is still highlighted as plain red.

### Expected behavior

Eventual line number highlighting should be a combination of `number_hl_number` and any relevant `*LineNr*`. Meaning that any highlight attribute should be taken from `number_hl_group` *with fallback to any active `*LineNr*`* .

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2400+g8952a89db

### Vim (not Nvim) behaves the same?

No (doesn't have this functionality)

### Operating system/version

EndeavourOS Linux x86_64 (6.7.5-arch1-1)

### Terminal name/version

st-0.9

### $TERM environment variable

st-256color

### Installation

appimage

**Comment by luukvbaal** [comment]
> ### Vim (not Nvim) behaves the same?
> 
> No (doesn't have this functionality)

Not true, `number_hl_group` just implements the legacy sign `numhl` property. FWIW, vim behaves the same, not combining any attrs for `numhl`: https://github.com/vim/vim/blob/15935e7f54f0e00782a55ebc39a68e4cd94c2571/src/drawline.c#L479

Whether it should, or we want to regardless of what vim does, is up for debate I suppose.

**Comment by echasnovski** [comment]
> Not true, `number_hl_group` just implements the legacy sign `numhl` property. FWIW, vim behaves the same, not combining any attrs for `numhl`: https://github.com/vim/vim/blob/15935e7f54f0e00782a55ebc39a68e4cd94c2571/src/drawline.c#L479

I see. Although, it looks more like an implementation detail, as extmarks are not the same as signs. As Vim does not have extmarks, to me it can not behave the same when question is about their behavior.

**Comment by luukvbaal** [comment]
To the contrary, extmark signs and legacy signs use the same code path and are functionally identical. The same "issue" can be reproduced in both vim and nvim with `:sign` commands.

**Comment by echasnovski** [comment]
> To the contrary, extmark signs and legacy signs use the same drawing code path and are functionally identical.

This is implementation details. As is the fact that `number_hl_group` is implemented using legacy signs.

Here is the documentation of `number_hl_group` option which says nothing about it being related to signs:
```
• number_hl_group: name of the highlight group used to
highlight the number column. Note: ranges are unsupported
and decorations are only applied to start_row
```

**Comment by luukvbaal** [comment]
I mean yeah but it's a relevant implementation detail. I don't think it makes sense to change one behavior without changing the other.

**Comment by luukvbaal** [comment]
The `:h sign-define` docs specifically state that `numhl` overrides the `*LineNr*` hl groups BTW.

**Comment by luukvbaal** [comment]
> Here is the documentation of `number_hl_group` option which says nothing about it being related to signs:

I think the extmark `*hl_group`s should mention their `sign` counterparts.

**Comment by bfredl** [comment]
overriding is an anti-pattern. Combining (with well-definied ordering) is how we process highlight sources in general, and should be how `number_hl_group` takes effect as well.

Still I concur the the _first_ thing to try is to send a patch/issue to vim, so we can keep the code as unified as possible while being consistent with vim, as long as this is posside.

**Comment by bfredl** [comment]
@zeertzjq Very simple: you just don't. if information source X adds a color and information source Y adds a bold attribute, the effect of having both sources at the same place is to have both the color and the attribute. If this is undesired, you reconfigure the source whose behavior you don't want. This is how highlight combining works all over the place. There is no reason to make `number_hl_group` different.

**Comment by zeertzjq** [comment]
> I don't think combining is a good idea. Removing attributes is harder than adding. E.g., how are you supposed to remove a `bold` or `italic` attribute when combining?

Oh, I forgot that `nocombine` can be used for that.

**Comment by luukvbaal** [comment]
Resolved by #31575, #32060, #32400.
